fix: keep autodoc type alias docstrings

casey-brooks opened the pull request:
## Summary
- handle ast.Constant string literals when picking up variable docstrings
- add regression fixtures exercising alias docstrings under simulated Python 3.14 behavior
- cover behaviour with a new autodoc test that monkeypatches ast.Str and asserts docstrings are emitted

## Testing
- PYTHONPATH=/workspace/sphinx .venv/bin/pytest tests/test_ext_autodoc_autodata.py -k type_alias_docstring -q
- PYTHONPATH=/workspace/sphinx .venv/bin/flake8 sphinx/pycode/parser.py tests/test_ext_autodoc_autodata.py tests/roots/test-ext-autodoc-typealias

Resolves #71

casey-brooks committed:
fix(autodoc): restore type alias docstrings

casey-brooks commented:
### Test & Lint Summary
- PYTHONPATH=/workspace/sphinx .venv/bin/pytest tests/test_ext_autodoc_autodata.py -k type_alias_docstring -q *(1 passed, 0 failed, 8 deselected)*
- .venv/bin/flake8 sphinx/pycode/parser.py tests/test_ext_autodoc_autodata.py tests/roots/test-ext-autodoc-typealias *(no issues)*

noa-lucent commented on tests/roots/test-ext-autodoc-typealias/type_alias_docstrings.py:
```diff
@@ -0,0 +1,38 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Any, Callable, Dict, TypeAlias
```
[major] Importing  from  will fail on the Python versions this branch still supports ( declares ). The new fixture gets imported while building the docs in the testroot, so an  on 3.6–3.9 would break the test suite (and users on those interpreters). Please either avoid using  here or provide a compatibility fallback (e.g. import from  when needed).

noa-lucent replied:
(superseded by the following comment—see thread PRRT_kwDOQE1Ros5naB6a for the actual feedback on this line).

noa-lucent commented on tests/roots/test-ext-autodoc-typealias/type_alias_docstrings.py:
```diff
@@ -0,0 +1,38 @@
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Any, Callable, Dict, TypeAlias
```
[major] Importing `TypeAlias` from `typing` will crash on the Python versions this branch still supports (`setup.py` still advertises `python_requires=\">=3.6\"`). This module is imported by the autodoc testroot, so on 3.6–3.9 the builder would raise an `ImportError` before we even exercise the regression. Please avoid the unconditional `TypeAlias` import (it is not needed for the reproduction), or add a compatibility fallback via `typing_extensions`.

noa-lucent commented on tests/roots/test-ext-autodoc-typealias/type_alias_docstrings.py:
```diff
@@ -0,0 +1,38 @@
+Similarly when **pretend** is ``True``, no operation should be really
+performed, but any action should be logged as if realized.
+"""
+
```
[major] `str | None` (and the similar union in `FileOp`) relies on PEP 604, which only works on Python 3.10+. On 3.6–3.9 this assignment either fails to parse or raises a `TypeError` because `type.__or__` is not defined. The test root is imported during the build, so we would regress all supported pre-3.10 interpreters. Please express these aliases with `typing.Optional` / `typing.Union` instead.

noa-lucent commented on tests/roots/test-ext-autodoc-typealias/type_alias_docstrings.py:
```diff
@@ -0,0 +1,38 @@
+"""Fixtures for type alias docstring handling tests."""
+
+from __future__ import annotations
+
```
[major] `from __future__ import annotations` is only recognized starting in Python 3.7. Our support matrix for this branch still includes Python 3.6 (see `python_requires`), so importing this module would raise a `SyntaxError` before the tests even run. Once the PEP 604 unions are rewritten, the future import is no longer needed—please drop it to keep the fixture loadable on 3.6.

noa-lucent reviewed (changes requested):
Thanks for tackling the Constant/Str regression and adding coverage. I left blocking notes on the new fixtures: the module currently relies on Python 3.7+/3.10+ syntax (`from __future__ import annotations`, `TypeAlias`, and the PEP 604 unions). Because the branch still supports Python 3.6–3.9, the testroot would fail to import there. Please adjust the fixtures to stay compatible with the supported interpreter range.

casey-brooks committed:
test(autodoc): support py36 in alias fixtures

casey-brooks commented:
Updated the fixture to avoid Python 3.10+ syntax: dropped the future annotations import, swapped the  usage for standard  unions, and reran the focused autodoc test + flake8.

casey-brooks commented:
Updated the fixture to avoid Python 3.10+ syntax: dropped the future annotations import, swapped the `TypeAlias` usage for standard `Optional[...]` unions, and reran the focused autodoc test + flake8.

noa-lucent reviewed (approved):
Looks great—thanks for backporting the fixture to older interpreters.
